privacyidea/lib/tokens/emailtoken.py

```diff
@@ -101,7 +104,11 @@ def __init__(self, aToken):
 
     @property
     def _email_address(self):
-        return self.get_tokeninfo(self.EMAIL_ADDRESS_KEY)
+        if is_true(self.get_tokeninfo("dynamic_email")):
+            email = self.user.info.get(self.EMAIL_ADDRESS_KEY)
+        else:
+            email = self.get_tokeninfo(self.EMAIL_ADDRESS_KEY)
```

Comment:
Hm, do we want to check for an empty email here (or somewhere else?)
Current behavior is that we just send an email with an empty "To" line.

Reply:
Likewise, we get the following message if the user has no ``email`` in the LDAP mapping:
```
[2018-02-20 17:19:21,255][14447][140257274791104][WARNING][privacyidea.lib.tokens.emailtoken:270] The PIN was correct, but the EMail could not be sent: TypeError('sequence item 0: expected string, NoneType found',)
```
But I guess that's not *so* bad?

Reply:
Honestly I am not sure.

I am even not sure if we should check for both cases or only in the dynamic case.

Reply:
I see. Then, we should maybe only add a log warning? "Warning: The email is sent to an empty recipient", "Warning: The SMS is sent to an empty phone number" or so.

Reply:
Doesn't  the lower level raise an exception?

Well, if we add code, then I think I would like to raise an exception myself. But the exception would be displayed to the user. :-/

So, a log warning might be better. Because this nags the admin with his log monitoring...

Reply:
Oh, I added a log warning. Bummer. THis adds two lines without code coverage.
Working on the test case...

Reply:
I just checked: In case of the LDAP resolver, if the LDAP user entry has no ``mail`` attribute, but the mapping contains a mapping to the ``email`` userinfo key, then no exception is thrown, the email recipient is just empty.
In case the mapping does not contain a mapping to the ``email`` userinfo key, the above exception is thrown.

Reply:
So everything is fine - isn't it?

If the administrator enrolls dynamic email tokens without an email mapping in his resolver config, he can come to me in person and I am happy to argue, if this is a wise idea or not! ;)

Shit in - shit out. Or do I fail to get the point?

Reply:
Yep, looks good now :) Guess I didn't see your new commit in time.